docs/proposals/cluster-capacity.md

```diff
@@ -0,0 +1,461 @@
+We need to know how many half CPU pods can be placed given the physical resources available
+on the nodes that match the node selector of the namespace.
+We also need to know how many of the largest pod definitions can be scaled up
+on this namespace (in this case, the half CPU pod can be scaled up to 8,
+assuming it's the only one defined).
+
+**Use Case 3**: A POD definition has a PV claim.
+We need to know how many nodes have available attachments such that
+the POD can be scheduled to nodes that fit the namespace node selector.
+
+### Autoscaling
+
+Introduction of new nodes or increase of node resources is not the only way how to deal with insufficient cluster capacity.
+One can take advantage of Horizontal Pod Autoscaling [1] scaling a number of replicas (Replication Controller, Deployment, ReplicaSet) based on the current resource utilization.
+Or Cluster Autoscaler [1].
+Still, the autoscaling is another indicator of insufficient resources which can result in node scaling.
+
+## Design considerations and design space
+
+Assuming the following:
+
+* predicting remaining cluster capacity on pod bases (e.g. "How many pods of a given shape I can schedule")
+* using the scheduler workflow as it is (applying the same predicates and priority functions/executing the same code wherever possible, using the same configuration)
+* prediction must not change the state of the cluster
+* provide general framework usable for any scheduling algorithm
+
+Scheduling process uses two sets of functions: predicates and priority functions.
+Predicates select all nodes suitable for scheduling.
+Priority functions choose the most suitable node from the selected nodes.
+Both predicates and priority functions may need to query the current state of the cluster during execution.
+Queries can include node's free cpu and memory, volume claims, annotations, etc.
+Once a pod gets scheduled on a node, it can change its free resources and change the decision basis of the next scheduling iteration.
+
+Thus, when designing the predictor, one has to take into account the current cluster state and compute its next without changing the current one.
+I.e. simulate the pod scheduling and binding. The simulation consists of (not exclusively in this order):
+
+* estimation of the current cluster state (node info, available volume claims, etc.)
+* execution of one scheduling iteration
+* update of scheduler caches based on scheduled pod
+
+Some consumed resources do not have to be specified in the pod (e.g. volume claims).
+These kind of consumed resources must be simulated accordingaly.
+
+Once the current state is captured, prediction is no longer dependent on the cluster state.
+Thus, the estimation does not have to correspond to the real scheduling since:
+
+* some nodes can be deleted, some nodes can be evacuated, some nodes can be re-labeled/re-annotated
+* some pods can get evicted in the time of prediction
```

Comment:
I would almost replace the word prediction with analysis is most places.    An analysis can be date stamped as of: x/y/z a:b:c  

statistics..